add DispatchKeySet function to get highest backend key

bdhirsh opened the pull request:
Adds a `c10::highestPriorityBackendTypeId(DispatchKeySet)` API that returns the highest priority *backend* dispatch key. A version of this used to exist before my recent dispatch key changes, but I deleted it because it wasn't actually needed anywhere in core. This is useful to mobile's static dispatch use case (cc @larryliu0820, @priyaramani).

This is similar to `getBackendKey()`, but `getBackendKey()` only returns a `BackendComponent`, which means that it can't distinguish between dense/sparse/quantized backends.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #75233

bdhirsh committed:
add DispatchKeySet function to get highest backend key

[ghstack-poisoned]

ezyang commented on c10/core/DispatchKeySet.h:
```diff
@@ -804,6 +814,14 @@ inline DispatchKeySet getAutocastRelatedKeySetFromBackend(BackendComponent t) {
+// This is basically like highestBackendKey(), except that we have some
+// "functionality" bits that correspond to backends (Sparse, Quantized)
+inline DispatchKey highestPriorityBackendTypeId(DispatchKeySet ks) {
+  auto backend_key = (ks & backend_functionality_keys).highestPriorityTypeId();
```
nit: can't you just return this directly?

larryliu0820 reviewed (approved):
Thanks for adding this!

bdhirsh committed:
Update on "add DispatchKeySet function to get highest backend key"

Adds a `c10::highestPriorityBackendTypeId(DispatchKeySet)` API that returns the highest priority *backend* dispatch key. A version of this used to exist before my recent dispatch key changes, but I deleted it because it wasn't actually needed anywhere in core. This is useful to mobile's static dispatch use case (cc @larryliu0820, @priyaramani).

This is similar to `getBackendKey()`, but `getBackendKey()` only returns a `BackendComponent`, which means that it can't distinguish between dense/sparse/quantized backends.




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge this please
